### ISSUE TITLE
Template variable unbound in rewrite rule

### ISSUE THREAD

**Issue by jacobstanley** [ekmett/hybrid-vectors#7]
This is a long shot, as I haven't had the time to narrow this down to a smaller example, but in some situations when I'm using `Hybrid.fromList` on GHC 7.10.2 (7.8.4 seems fine) I get a compiler error like the following:

``` haskell
<no location info>:
    ghc: panic! (the 'impossible' happened)
  (GHC version 7.10.2 for x86_64-unknown-linux):
    Template variable unbound in rewrite rule
  sc_sJxx
  [sc_sJxw, sc_sJxx, sc_sJxy, sc_sJxz, sg_sJxA]
  [sc_sJxw, sc_sJxx, sc_sJxy, sc_sJxz, sg_sJxA]
  [sc_sJxw,
   (MV
      @ (Mutable Vector)
      @ (Mutable Vector)
      @ (PrimState (ST s2_id4I))
      @ (ByteString, Int64)
      @ ByteString
      @ Int64
      @~ <(ByteString, Int64)>_N
      sc_sJxy
      sc_sJxz)
   `cast` (sg_sJxA
           :: MVector
                (Mutable Vector)
                (Mutable Vector)
                (PrimState (ST s2_id4I))
                (ByteString, Int64)
              ~R# Mutable
                    (Vector Vector Vector)
                    (PrimState (ST s2_id4I))
                    (ByteString, Int64))]
  [sc_sJxq,
   wild_idn0
   `cast` (Sub
             (Sym (TFCo:R:MutableVector[0] <Vector>_N <Vector>_N)) <PrimState
                                                                      (ST s2_id4I)>_N <(ByteString,
                                                                                        Int64)>_N
           :: MVector
                (Mutable Vector)
                (Mutable Vector)
                (PrimState (ST s2_id4I))
                (ByteString, Int64)
              ~R# Mutable
                    (Vector Vector Vector)
                    (PrimState (ST s2_id4I))
                    (ByteString, Int64))]

Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug
```

If you have any ideas/clues as to why this might be happening then that would be great, if not then feel free to close this issue and I'll try to narrow it down a bit more.

I understand this is probably a GHC bug, as it says in the message, but was wondering if there might be something that can be done to work around it.

**Comment by RyanGlScott** [comment]
Can you still reproduce this with GHC 8.0?

**Comment by jacobstanley** [comment]
> Can you still reproduce this with GHC 8.0?

I'll try to get that failing case set up again when I get a chance, but fair warning this isn't a priority for me at the moment, so it could be a little while.

**Comment by RyanGlScott** [comment]
Funny enough, I managed to reproduce this very error on GHC HEAD (8.3.20170310). I've opened [GHC Trac #13410](https://ghc.haskell.org/trac/ghc/ticket/13410) for this.

**Comment by RyanGlScott** [comment]
GHC Trac #13410 was just fixed, and it should be merged into GHC 8.2.

**Comment by jacobstanley** [comment]
\o/
